File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupPollsListSheet.kt
package com.example.momentra.ui.shell.group.shared

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.data.api.GroupPollItemDto
import com.example.momentra.ui.theme.PlusJakartaSans
import java.util.Locale
import kotlin.math.roundToInt

private enum class PollsListFilter(val label: String) {
    All("All"),
    Active("Active"),
    Closed("Closed"),
}

private val pollAvatarColors = listOf(
    Color(0xFFFDBA74),
    Color(0xFF86EFAC),
    Color(0xFF93C5FD),
    Color(0xFFC4B5FD),
)

/** Figma 1604:16811 — Polls list (View all). Live APIs only. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GroupPollsListSheet(
    visible: Boolean,
    momentTitle: String?,
    chrome: MomentsChrome,
    polls: List<GroupPollItemDto>,
    onDismiss: () -> Unit,
    onChanged: () -> Unit = {},
) {
    if (!visible) return
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = chrome.bg,
    ) {
        GroupPollsListBody(
            momentTitle = momentTitle,
            chrome = chrome,
            polls = polls,
            onDismiss = onDismiss,
            onChanged = onChanged,
        )
    }
}

@Composable
fun GroupPollsListBody(
    momentTitle: String?,
    chrome: MomentsChrome,
    polls: List<GroupPollItemDto>,
    onDismiss: () -> Unit,
    onChanged: () -> Unit = {},
) {
    var filter by remember { mutableStateOf(PollsListFilter.All) }
    var selectedPollId by remember { mutableStateOf<String?>(null) }

    val filtered = remember(polls, filter) {
        when (filter) {
            PollsListFilter.All -> polls
            PollsListFilter.Active -> polls.filter { it.status.equals("OPEN", ignoreCase = true) }
            PollsListFilter.Closed -> polls.filter {
                val s = it.status.orEmpty().uppercase(Locale.US)
                s == "CLOSED" || s == "CANCELLED"
            }
        }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .fillMaxHeight(0.95f)
            .background(chrome.bg),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 10.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(CircleShape)
                    .background(chrome.card)
                    .border(1.dp, chrome.border, CircleShape)
                    .clickable(onClick = onDismiss),
                contentAlignment = Alignment.Center,
            ) {
                Text("←", color = chrome.text, fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
            }
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    "Polls",
                    color = chrome.text,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = PlusJakartaSans,
                )
                Text(
                    momentTitle ?: "Shared moment",
                    color = chrome.secondary,
                    fontSize = 12.sp,
                    fontFamily = PlusJakartaSans,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            PollsListFilter.entries.forEach { tab ->
                val selected = filter == tab
                Text(
                    tab.label,
                    color = if (selected) chrome.darkText else chrome.secondary,
                    fontSize = 13.sp,
                    fontWeight = if (selected) FontWeight.Bold else FontWeight.SemiBold,
                    fontFamily = PlusJakartaSans,
                    modifier = Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(if (selected) chrome.accent else chrome.card)
                        .then(
                            if (selected) Modifier else Modifier.border(1.dp, chrome.border, RoundedCornerShape(999.dp)),
                        )
                        .clickable { filter = tab }
                        .padding(horizontal = 16.dp, vertical = 8.dp),
                )
            }
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp, vertical = 12.dp)
                .padding(bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            if (filtered.isEmpty()) {
                val (title, detail) = when (filter) {
                    PollsListFilter.All -> "No polls yet" to "Create a poll from Quick Add — nothing is invented."
                    PollsListFilter.Active -> "No active polls" to "Open polls will show here."
                    PollsListFilter.Closed -> "No closed polls" to "Closed polls will show here."
                }
                GroupEmptySection(title, detail)
            } else {
                filtered.forEachIndexed { index, poll ->
                    PollsListCard(
                        poll = poll,
                        index = index,
                        chrome = chrome,
                        onVote = { poll.pollId?.let { selectedPollId = it } },
                        onOpen = { poll.pollId?.let { selectedPollId = it } },
                    )
                }
            }
        }
    }

    selectedPollId?.let { pollId ->
        PollDetailSheet(
            pollId = pollId,
            visible = true,
            onDismiss = { selectedPollId = null },
            onSaved = {
                onChanged()
            },
        )
    }
}

@Composable
private fun PollsListCard(
    poll: GroupPollItemDto,
    index: Int,
    chrome: MomentsChrome,
    onVote: () -> Unit,
    onOpen: () -> Unit,
) {
    val isOpen = poll.status.equals("OPEN", ignoreCase = true)
    val options = poll.options
    val total = maxOf(poll.totalVotes ?: options.sumOf { it.voteCount ?: 0 }, 0)
    val denom = maxOf(total, 1)
    val creator = poll.createdByDisplayName ?: "Member"
    val endsTag = formatPollEndsTag(poll.closesAt, poll.status)
    val winningCount = options.maxOfOrNull { it.voteCount ?: 0 } ?: 0

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(chrome.card)
            .border(1.dp, chrome.border, RoundedCornerShape(16.dp))
            .clickable(onClick = onOpen)
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.weight(1f, fill = false),
            ) {
                Box(
                    modifier = Modifier
                        .size(24.dp)
                        .clip(CircleShape)
                        .background(pollAvatarColors[index % pollAvatarColors.size]),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        initialsFromName(creator),
                        color = chrome.darkText,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = PlusJakartaSans,
                    )
                }
                Text(
                    creator,
                    color = chrome.secondary,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    fontFamily = PlusJakartaSans,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            Text(
                endsTag,
                color = if (isOpen) chrome.accent else chrome.secondary,
                fontSize = 10.sp,
                fontWeight = FontWeight.Bold,
                fontFamily = PlusJakartaSans,
                modifier = Modifier
                    .clip(RoundedCornerShape(6.dp))
                    .background(if (isOpen) chrome.brandSoft else chrome.card)
                    .then(
                        if (isOpen) Modifier else Modifier.border(1.dp, chrome.border, RoundedCornerShape(6.dp)),
                    )
                    .padding(horizontal = 8.dp, vertical = 4.dp),
            )
        }

        Text(
            poll.question ?: "Poll",
            color = chrome.text,
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = PlusJakartaSans,
        )

        Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
            options.forEach { option ->
                val votes = option.voteCount ?: 0
                val pct = ((votes.toDouble() / denom) * 100).roundToInt()
                val isWinner = !isOpen && winningCount > 0 && votes == winningCount
                Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                            if (isWinner) {
                                Text("✓", color = chrome.accent, fontSize = 12.sp, fontWeight = FontWeight.Bold)
                            }
                            Text(
                                option.text ?: "Option",
                                color = chrome.text,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.SemiBold,
                                fontFamily = PlusJakartaSans,
                            )
                        }
                        Text(
                            "$votes votes ($pct%)",
                            color = chrome.secondary,
                            fontSize = 12.sp,
                            fontFamily = PlusJakartaSans,
                        )
                    }
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(8.dp)
                            .clip(RoundedCornerShape(999.dp))
                            .background(Color(0xFF252332)),
                    ) {
                        val fraction = if (total == 0) 0f else (votes.toFloat() / denom).coerceIn(0f, 1f)
                        Box(
                            modifier = Modifier
                                .fillMaxWidth(fraction.coerceAtLeast(if (total == 0) 0f else 0.02f))
                                .height(8.dp)
                                .clip(RoundedCornerShape(999.dp))
                                .background(chrome.accent),
                        )
                    }
                }
            }
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                if (total == 1) "1 total vote" else "$total total votes",
                color = chrome.secondary,
                fontSize = 12.sp,
                fontFamily = PlusJakartaSans,
            )
            if (isOpen) {
                Text(
                    "Vote",
                    color = chrome.darkText,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = PlusJakartaSans,
                    modifier = Modifier
                        .clip(RoundedCornerShape(8.dp))
                        .background(chrome.accent)
                        .clickable(onClick = onVote)
                        .padding(horizontal = 16.dp, vertical = 6.dp),
                )
            }
        }
    }
}

File: momentra/momentra/Shell/GroupActive/Shared/GroupPollsListSheet.swift
import SwiftUI

/// Figma 1604:16811 — Polls list (View all). Live APIs only.
struct GroupPollsListSheet: View {
    let momentTitle: String?
    var chrome: MomentsChrome = .trip
    let polls: [APIClient.GroupPollItemPayload]
    var onDismiss: () -> Void
    var onChanged: () -> Void = {}

    @State private var filter: PollsListFilter = .all
    @State private var selectedPollId: String?

    private enum PollsListFilter: String, CaseIterable, Identifiable {
        case all = "All"
        case active = "Active"
        case closed = "Closed"
        var id: String { rawValue }
    }

    private var avatarColors: [Color] {
        [Color(hex: "#FDBA74"), Color(hex: "#86EFAC"), Color(hex: "#93C5FD"), Color(hex: "#C4B5FD")]
    }

    private var filtered: [APIClient.GroupPollItemPayload] {
        switch filter {
        case .all: return polls
        case .active:
            return polls.filter { ($0.status ?? "").uppercased() == "OPEN" }
        case .closed:
            return polls.filter {
                let s = ($0.status ?? "").uppercased()
                return s == "CLOSED" || s == "CANCELLED"
            }
        }
    }

    var body: some View {
        VStack(spacing: 0) {
            header
            filterTabs
            ScrollView {
                LazyVStack(spacing: 16) {
                    if filtered.isEmpty {
                        GroupEmptySection(
                            message: emptyTitle,
                            detail: emptyDetail
                        )
                        .padding(.top, 24)
                    } else {
                        ForEach(Array(filtered.enumerated()), id: \.element.id) { index, poll in
                            pollCard(poll, index: index)
                        }
                    }
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 12)
                .padding(.bottom, 32)
            }
        }
        .background(chrome.bg.ignoresSafeArea())
        .sheet(item: Binding(
            get: { selectedPollId.map { PollSheetItem(id: $0) } },
            set: { selectedPollId = $0?.id }
        )) { item in
            PollDetailSheet(
                pollId: item.id,
                onDismiss: { selectedPollId = nil },
                onSaved: { onChanged() }
            )
        }
    }

    private struct PollSheetItem: Identifiable {
        let id: String
    }

    private var emptyTitle: String {
        switch filter {
        case .all: return "No polls yet"
        case .active: return "No active polls"
        case .closed: return "No closed polls"
        }
    }

    private var emptyDetail: String {
        switch filter {
        case .all: return "Create a poll from Quick Add — nothing is invented."
        case .active: return "Open polls will show here."
        case .closed: return "Closed polls will show here."
        }
    }

    private var header: some View {
        HStack(spacing: 12) {
            Button(action: onDismiss) {
                Image(systemName: "chevron.left")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(chrome.text)
                    .frame(width: 36, height: 36)
                    .background(chrome.card)
                    .clipShape(Circle())
                    .overlay(Circle().stroke(chrome.border))
            }
            .buttonStyle(.plain)
            VStack(alignment: .leading, spacing: 2) {
                Text("Polls")
                    .font(.plusJakarta(size: 18, weight: .bold))
                    .foregroundStyle(chrome.text)
                Text(momentTitle ?? "Shared moment")
                    .font(.plusJakarta(size: 12))
                    .foregroundStyle(chrome.secondary)
                    .lineLimit(1)
            }
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 10)
    }

    private var filterTabs: some View {
        HStack(spacing: 8) {
            ForEach(PollsListFilter.allCases) { tab in
                let selected = filter == tab
                Button {
                    filter = tab
                } label: {
                    Text(tab.rawValue)
                        .font(.plusJakarta(size: 13, weight: selected ? .bold : .semibold))
                        .foregroundStyle(selected ? chrome.darkText : chrome.secondary)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 8)
                        .background(selected ? chrome.accent : chrome.card)
                        .clipShape(Capsule())
                        .overlay(
                            Capsule().stroke(selected ? Color.clear : chrome.border)
                        )
                }
                .buttonStyle(.plain)
            }
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }

    @ViewBuilder
    private func pollCard(_ poll: APIClient.GroupPollItemPayload, index: Int) -> some View {
        let isOpen = (poll.status ?? "").uppercased() == "OPEN"
        let options = poll.options ?? []
        let total = max(poll.totalVotes ?? options.reduce(0) { $0 + ($1.voteCount ?? 0) }, 0)
        let totalDenom = max(total, 1)
        let creator = poll.createdByDisplayName ?? "Member"
        let endsTag = formatPollEndsTag(closesAt: poll.closesAt, status: poll.status)
        let winningCount = options.map { $0.voteCount ?? 0 }.max() ?? 0

        VStack(alignment: .leading, spacing: 14) {
            HStack {
                HStack(spacing: 8) {
                    Text(initialsFromName(creator))
                        .font(.plusJakarta(size: 10, weight: .bold))
                        .foregroundStyle(chrome.darkText)
                        .frame(width: 24, height: 24)
                        .background(avatarColors[index % avatarColors.count])
                        .clipShape(Circle())
                    Text(creator)
                        .font(.plusJakarta(size: 12, weight: .semibold))
                        .foregroundStyle(chrome.secondary)
                        .lineLimit(1)
                }
                Spacer(minLength: 8)
                Text(endsTag)
                    .font(.plusJakarta(size: 10, weight: .bold))
                    .foregroundStyle(isOpen ? chrome.accent : chrome.secondary)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(isOpen ? chrome.brandSoft : chrome.card)
                    .clipShape(RoundedRectangle(cornerRadius: 6))
                    .overlay(
                        RoundedRectangle(cornerRadius: 6)
                            .stroke(isOpen ? Color.clear : chrome.border)
                    )
            }

            Text(poll.question ?? "Poll")
                .font(.plusJakarta(size: 16, weight: .bold))
                .foregroundStyle(chrome.text)
                .fixedSize(horizontal: false, vertical: true)

            VStack(alignment: .leading, spacing: 12) {
                ForEach(Array(options.enumerated()), id: \.offset) { _, option in
                    let votes = option.voteCount ?? 0
                    let pct = Int((Double(votes) / Double(totalDenom) * 100).rounded())
                    let isWinner = !isOpen && winningCount > 0 && votes == winningCount
                    VStack(alignment: .leading, spacing: 6) {
                        HStack(alignment: .firstTextBaseline) {
                            HStack(spacing: 6) {
                                if isWinner {
                                    Image(systemName: "checkmark.circle.fill")
                                        .font(.system(size: 12))
                                        .foregroundStyle(chrome.accent)
                                }
                                Text(option.text ?? "Option")
                                    .font(.plusJakarta(size: 13, weight: .semibold))
                                    .foregroundStyle(chrome.text)
                            }
                            Spacer(minLength: 8)
                            Text("\(votes) votes (\(pct)%)")
                                .font(.plusJakarta(size: 12))
                                .foregroundStyle(chrome.secondary)
                        }
                        GeometryReader { geo in
                            ZStack(alignment: .leading) {
                                Capsule().fill(Color(hex: "#252332"))
                                Capsule()
                                    .fill(chrome.accent)
                                    .frame(width: max(total == 0 ? 0 : 6, geo.size.width * CGFloat(votes) / CGFloat(totalDenom)))
                            }
                        }
                        .frame(height: 8)
                    }
                }
            }

            HStack {
                Text(total == 1 ? "1 total vote" : "\(total) total votes")
                    .font(.plusJakarta(size: 12))
                    .foregroundStyle(chrome.secondary)
                Spacer()
                if isOpen {
                    Button {
                        if let id = poll.pollId { selectedPollId = id }
                    } label: {
                        Text("Vote")
                            .font(.plusJakarta(size: 13, weight: .bold))
                            .foregroundStyle(chrome.darkText)
                            .padding(.horizontal, 16)
                            .padding(.vertical, 6)
                            .background(chrome.accent)
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(chrome.card)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .overlay(RoundedRectangle(cornerRadius: 16).stroke(chrome.border))
        .contentShape(RoundedRectangle(cornerRadius: 16))
        .onTapGesture {
            if let id = poll.pollId { selectedPollId = id }
        }
    }
}
